kms/managed.go

```diff
@@ -0,0 +1,224 @@
+
+	return &mk, nil
+}
+
+func (m *ManagedCipher) manageKey() error {
+	// Expensive lock that stops the world until we have a new key set up.
+	m.m.Lock()
+	defer m.m.Unlock()
+
+	// If the key is safe to use, do nothing.
+	m.counter = m.counter + 1
+	if m.counter < CollisionTolerance && time.Now().Before(m.expires) {
+		return nil
+	}
+
+	// If the key is not so safe to use, create a new key and setup the cipher
+	// anew.
+	cipher, err := NewAESGCM()
+	if err != nil {
+		klog.Infof("create new cipher: %w", err)
+		return err
+	}
+
+	// Let upstream KMS encrypt the new key in use.
+	keyID, encKey, err := m.upstreamCipher.Encrypt(cipher.Key())
+	if err != nil {
+		klog.Infof("encrypt with upstream: %w", err)
+		return err
+	}
+
+	// Update the current state
+	m.keys.Add(encKey, cipher)
+	m.currentLocalKEK = encKey
+	m.remoteKMSID = keyID
+	m.expires = time.Now().Add(week)
+	m.counter = 0
+
+	klog.Infof("created key successfully and added to set")
+
+	return nil
+}
+
+// Encrypt encrypts given plaintext and returns the key used in encrypted form.
+// The encrypted key is encrypted by the given upstream KMS.
+func (m *ManagedCipher) Encrypt(pt []byte) ([]byte, []byte, []byte, error) {
+	if err := m.manageKey(); err != nil {
+		return nil, nil, nil, fmt.Errorf("manage keys upfront of an encryption: %w", err)
+	}
+
+	cipher, ok := m.keys.Get(m.currentLocalKEK)
+	if !ok {
+		klog.Infof(
+			"current plugin key (%q) has no value in cache",
+			base64.StdEncoding.EncodeToString(m.currentLocalKEK),
+		)
+		return nil, nil, nil, fmt.Errorf(
+			"plugin is broken, current key is unknown (%q)",
+			base64.StdEncoding.EncodeToString(m.currentLocalKEK),
+		)
+	}
+
+	ct, err := cipher.Encrypt(pt)
+	if err != nil {
+		klog.Infof("encrypt plaintext: %w", err)
+		return nil, nil, nil, err
+	}
+
+	return m.remoteKMSID, m.currentLocalKEK, ct, nil
+}
+
+// DecryptRemotely decrypts given ciphertext by sendin it directly to the
+// remote kms.
+func (m *ManagedCipher) DecryptRemotely(id, ct []byte) ([]byte, error) {
+	return m.upstreamCipher.Decrypt(id, ct)
+}
+
+// Decrypt decrypts the given ciphertext. If the given encrypted key is unknown,
+// KMS upstream is asked for decryption of the encrypted key.
+func (m *ManagedCipher) Decrypt(keyID, encKey, ct []byte) ([]byte, error) {
+	// Lookup key from cache.
+	cipher, ok := m.keys.Get(encKey)
+	if ok {
+		pt, err := cipher.Decrypt(ct)
+		if err != nil {
+			klog.Infof("decrypt ciphertext: %w", err)
+			return nil, err
+		}
+
+		return pt, nil
+	}
+
+	klog.Infof(
+		"key (%q) has no value in cache",
+		base64.StdEncoding.EncodeToString(encKey),
+	)
+
+	// plainKey is a plaintext key and should be handled cautiously.
+	plainKey, err := m.upstreamCipher.Decrypt(keyID, encKey)
+	if err != nil {
+		klog.Infof(
+			"decrypt key (%q) by upstream:",
+			base64.StdEncoding.EncodeToString(encKey),
+			err,
+		)
+
+		return nil, err
+	}
+
+	// Assume that keyID is the most up to date upstream KMS key.
```

Comment:
Yes, this one is tricky. Do we assume that new is always better? Is there a better strategy?

Reply:
Removed after hint from @aramase